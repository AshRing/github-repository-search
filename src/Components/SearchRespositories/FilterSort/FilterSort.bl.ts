export const sortInputs = {
	bestMatch: "Best Match",
	stars: "Stars",
};

export const availableSorts = Object.keys(sortInputs).map((key: string) => sortInputs[key]);

export const sortValues = {
	[sortInputs.bestMatch]: undefined,
	[sortInputs.stars]: "stars",
};

export const filters = {
	language: "language",
};

export const availableFilters = Object.keys(filters).map((key: string) => filters[key]);

export const languageFilterOptions = [
	"Assembly",
	"C",
	"C#",
	"C++",
	"Dart",
	"HTML",
	"Haskell",
	"Go",
	"Java",
	"JavaScript",
	"Python",
	"Ruby",
	"Rust",
	"Shell",
	"TypeScript",
];
