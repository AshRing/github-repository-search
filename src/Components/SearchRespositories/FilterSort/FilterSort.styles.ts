import styled from "styled-components";

export const SearchInputContainer = styled.div`
	position: relative;
	width: 100%;
	margin-bottom: 1rem;

	& > .fa-search {
		position: absolute;
		top: 50%;
		left: 1rem;
		transform: translate(-50%, -50%);
		color: ${({ theme }) => theme.colors.secondary};
	}
`;

export const SearchInput = styled.input`
	width: 100%;
	padding: 0 1rem 0 2rem;
	height: 2.5rem;
	border-radius: 0.25rem;
	border: 1px solid ${({ theme }) => theme.colors.secondary};
	outline: none;
	transition: 0.2s ease border-color;

	&:focus {
		border-color: ${({ theme }) => theme.colors.primary};
	}
`;

export const FilterSortContainer = styled.div`
	display: flex;
	column-gap: 1rem;

	& > div {
		width: 50%;
	}
`;
