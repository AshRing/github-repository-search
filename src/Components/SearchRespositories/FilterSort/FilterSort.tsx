import * as React from "react";

import {
	FilterSortContainer,
	FilterSortGroup,
	SearchInput,
	SearchInputContainer,
} from "./FilterSort.styles";
import { availableSorts, filters, languageFilterOptions } from ".";
import { faSearch, faTimes } from "@fortawesome/free-solid-svg-icons";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { IFilterSortOption } from "src/_types";
import { SelectInput } from "../../../Components/Shared";

interface Props {
	addFilter: (filterToAdd: IFilterSortOption) => void;
	changeSearchTerm: (value: string) => void;
	changeSort: (sortName: string) => void;
	filterBy: IFilterSortOption[];
	searchRepositories: () => void;
	searchTerm: string;
	sortBy: string;
}

export const FilterSort = ({
	addFilter,
	changeSearchTerm,
	changeSort,
	filterBy,
	searchRepositories,
	searchTerm,
	sortBy,
}: Props) => {
	const searchInputRef: React.RefObject<HTMLInputElement> = React.useRef();

	React.useEffect(() => {
		searchInputRef.current.focus();
	}, []);

	return (
		<FilterSortContainer
			onSubmit={(e) => {
				e.preventDefault();
				searchRepositories();
			}}
		>
			<SearchInputContainer>
				<FontAwesomeIcon icon={faSearch} />
				<SearchInput
					ref={searchInputRef}
					type="text"
					placeholder="Search Repositories"
					value={searchTerm}
					onChange={(e: React.ChangeEvent) => changeSearchTerm(e.target.value)}
				/>
				{searchTerm && (
					<button type="button" onClick={() => changeSearchTerm("")}>
						<FontAwesomeIcon icon={faTimes} />
					</button>
				)}
			</SearchInputContainer>
			<FilterSortGroup>
				<div>
					<SelectInput
						multiselect
						label="Language"
						selected={
							filterBy.find(
								(filter: IFilterSortOption) => filter.name === filters.language,
							)?.values || []
						}
						options={languageFilterOptions}
						handleChange={(selected: string[]) =>
							addFilter({ name: filters.language, values: selected })
						}
					/>
				</div>
				<div>
					<SelectInput
						label="Sort"
						selected={[sortBy]}
						options={availableSorts}
						handleChange={(selected: string[]) => changeSort(selected[0])}
					/>
				</div>
			</FilterSortGroup>
		</FilterSortContainer>
	);
};
