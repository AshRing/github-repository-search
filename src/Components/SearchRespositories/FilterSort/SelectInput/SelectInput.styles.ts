import { mediaQueries } from "../../../../_style";
import { rgba } from "polished";
import styled from "styled-components";

const inputHeight = "2.5rem";

export const SelectInputContainer = styled.div<{ inputActive: boolean }>`
	position: relative;
	height: ${inputHeight};
	border-radius: 0.25rem;
	padding: 0 2rem 0 1rem;
	transition: 0.2s ease border;
	display: flex;
	align-items: center;
	cursor: pointer;
	background: #fff;
	border: 1px solid
		${({ inputActive, theme }) =>
			!inputActive ? theme.colors.secondary : theme.colors.primary};
`;

export const SelectLabel = styled.span<{ shrinkLabel: boolean }>`
	position: absolute;
	top: ${({ shrinkLabel }) => (shrinkLabel ? "0" : "50%")};
	left: ${({ shrinkLabel }) => (shrinkLabel ? ".5rem" : "1rem")};
	font-size: ${({ shrinkLabel }) => (shrinkLabel ? ".75rem" : "1rem")};
	transform: translateY(-50%);
	transition: 0.2s ease all;
	padding: ${({ shrinkLabel }) => shrinkLabel && ".2rem"};
	color: ${({ theme }) => theme.colors.primary};
	background-color: #fff;
	background: ${({ shrinkLabel, theme }) =>
		shrinkLabel
			? `linear-gradient(to bottom, ${theme.colors.tertiary} 50%, #fff 50%)`
			: "#fff"};
`;

export const SelectedText = styled.span`
	width: 100%;
	white-space: nowrap;
	text-overflow: ellipsis;
	overflow: hidden;
`;

export const DrowdownArrow = styled.div`
	position: absolute;
	top: 50%;
	right: 0.5rem;
	transform: translateY(-50%);
	& > svg {
		color: ${({ theme }) => theme.colors.secondary};
	}

	.largeScreen {
		display: none;

		${mediaQueries.desktop} {
			display: initial;
		}
	}

	.smallScreen {
		${mediaQueries.desktop} {
			display: none;
		}
	}
`;

export const SelectDropdown = styled.div<{ open: boolean }>`
	position: fixed;
	top: 0;
	bottom: 0;
	right: ${({ open }) => (open ? "0" : "-100%")};
	width: 75%;
	background: ${({ theme }) => theme.colors.secondary};
	padding: 2rem;
	transition: 0.3s ease-out all;
	z-index: 5;

	${mediaQueries.tabletPortrait} {
		width: 50%;
	}

	${mediaQueries.desktop} {
		position: absolute;
		bottom: auto;
		left: 0;
		width: 100%;
		top: calc(${inputHeight} + 0.25rem);
		opacity: ${({ open }) => (open ? "1" : "0")};
		transform-origin: top;
		transform: scaleY(${({ open }) => (open ? "1" : "0")});
		border-radius: 0.25rem;
		background: #fff;
		border: 1px solid ${({ theme }) => theme.colors.secondary};
		box-shadow: 2px 2px 2px ${({ theme }) => rgba(theme.colors.primary, 0.3)};
		padding: 1rem 1.5rem;
	}

	& > h1 {
		font-size: 1.5rem;
		margin-bottom: 2.5rem;

		${mediaQueries.desktop} {
			display: none;
		}
	}

	& .fa-times {
		position: absolute;
		top: 2rem;
		right: 2rem;
		color: ${({ theme }) => theme.colors.primary};
		font-size: 2rem;

		${mediaQueries.desktop} {
			display: none;
		}
	}
`;

export const SelectOption = styled.div<{ selected: boolean }>`
	display: flex;

	&:not(:last-child) {
		margin-bottom: 0.5rem;
	}

	& > svg {
		margin-right: 0.5rem;
		color: ${({ theme }) => theme.colors.primary};
	}

	& > span {
		font-weight: ${({ selected }) => selected && "bold"};
	}
`;

export const ClearButton = styled.button`
	font-size: 1rem;
	font-weight: bold;
	margin: 1rem 0 0;
`;
