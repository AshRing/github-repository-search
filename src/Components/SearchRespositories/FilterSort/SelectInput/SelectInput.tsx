import * as React from "react";

import {
	ClearButton,
	DrowdownArrow,
	SelectDropdown,
	SelectInputContainer,
	SelectLabel,
	SelectOption,
	SelectedText,
} from "./SelectInput.styles";
import {
	faArrowDown,
	faArrowLeft,
	faArrowRight,
	faArrowUp,
	faCheckSquare,
	faSquare,
	faTimes,
} from "@fortawesome/free-solid-svg-icons";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import useOnClickOutside from "use-onclickoutside";

interface Props {
	handleChange: (selected: string[]) => void;
	label: string;
	multiselect?: boolean;
	options: string[];
	selected: string[];
}

export const SelectInput = ({ handleChange, label, multiselect, options, selected }: Props) => {
	const [dropdownOpen, toggleDropdown] = React.useState<boolean>(false);
	const containerRef: React.RefObject<HTMLDivElement> = React.useRef();
	const dropdownRef: React.RefObject<HTMLDivElement> = React.useRef();
	useOnClickOutside(
		dropdownRef,
		(e) => !containerRef.current.contains(e.target as Node) && toggleDropdown(false),
	);

	const handleMultiselectOptionSelection = (multiselectSelectedOption: string) => {
		if (selected.includes(multiselectSelectedOption)) {
			const optionRemoved: string[] = selected.filter(
				(selectedOption: string) => selectedOption !== multiselectSelectedOption,
			);
			handleChange(optionRemoved);
		} else {
			handleChange([...selected, multiselectSelectedOption]);
		}
	};

	const renderOptions = () => {
		return options.map((option: string, i: number) => {
			if (multiselect) {
				const optionSelected: boolean = selected.includes(option);

				return (
					<SelectOption
						key={`${option}-${i}`}
						selected={optionSelected}
						onClick={() => handleMultiselectOptionSelection(option)}
					>
						{optionSelected ? (
							<FontAwesomeIcon
								className="multiSelectOptionSelected"
								icon={faCheckSquare}
							/>
						) : (
							<FontAwesomeIcon
								className="multiSelectOptionNotSelected"
								icon={faSquare}
							/>
						)}
						<span>{option}</span>
					</SelectOption>
				);
			} else {
				const optionSelected: boolean = selected.includes(option);
				return (
					<SelectOption
						key={`${option}-${i}`}
						selected={optionSelected}
						onClick={() => {
							handleChange([option]);
							toggleDropdown(false);
						}}
					>
						<span>{option}</span>
					</SelectOption>
				);
			}
		});
	};

	return (
		<SelectInputContainer
			ref={containerRef}
			inputActive={selected.length || dropdownOpen}
			onClick={(e) =>
				!dropdownRef.current.contains(e.target) ? toggleDropdown(!dropdownOpen) : undefined
			}
		>
			<SelectLabel shrinkLabel={selected.length}>{label}</SelectLabel>
			<SelectedText>{selected.join(", ")}</SelectedText>
			<DrowdownArrow>
				{dropdownOpen ? (
					<>
						<FontAwesomeIcon
							icon={faArrowUp}
							className="dropdownOpenIcon largeScreen"
						/>
						<FontAwesomeIcon
							icon={faArrowRight}
							className="dropdownOpenIcon smallScreen"
						/>
					</>
				) : (
					<>
						<FontAwesomeIcon
							icon={faArrowDown}
							className="dropdownClosedIcon largeScreen"
						/>
						<FontAwesomeIcon
							icon={faArrowLeft}
							className="dropdownClosedIcon smallScreen"
						/>
					</>
				)}
			</DrowdownArrow>
			<SelectDropdown open={dropdownOpen} ref={dropdownRef}>
				<button id="dropdownCloseIcon" type="button" onClick={() => toggleDropdown(false)}>
					<FontAwesomeIcon icon={faTimes} />
				</button>
				<h1>{label}</h1>
				{renderOptions()}
				{selected.length > 0 && multiselect && (
					<ClearButton type="button" onClick={() => handleChange([])}>
						Clear
					</ClearButton>
				)}
			</SelectDropdown>
		</SelectInputContainer>
	);
};
