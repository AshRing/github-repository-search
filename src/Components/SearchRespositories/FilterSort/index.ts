export * from "./FilterSort";
export * from "./FilterSort.bl";
