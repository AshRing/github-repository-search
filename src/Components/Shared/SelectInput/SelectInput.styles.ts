import styled from "styled-components";

export const SelectInputContainer = styled.div<{ inputActive: boolean }>`
	position: relative;
	height: 2.5rem;
	border-radius: 0.25rem;
	padding: 0 2rem 0 1rem;
	transition: 0.2s ease border;
	display: flex;
	align-items: center;
	cursor: pointer;
	border: 1px solid
		${({ inputActive, theme }) =>
			!inputActive ? theme.colors.secondary : theme.colors.primary};
`;

export const SelectLabel = styled.span<{ shrinkLabel: boolean }>`
	position: absolute;
	top: ${({ shrinkLabel }) => (shrinkLabel ? "0" : "50%")};
	left: ${({ shrinkLabel }) => (shrinkLabel ? ".5rem" : "1rem")};
	font-size: ${({ shrinkLabel }) => (shrinkLabel ? ".75rem" : "1rem")};
	transform: translateY(-50%);
	transition: 0.2s ease all;
	padding: ${({ shrinkLabel }) => shrinkLabel && ".2rem"};
	color: ${({ theme }) => theme.colors.primary};
	background-color: #fff;
`;

export const SelectedText = styled.span`
	width: 100%;
	white-space: nowrap;
	text-overflow: ellipsis;
	overflow: hidden;
`;

export const DrowdownArrow = styled.div`
	position: absolute;
	top: 50%;
	right: 0.5rem;
	transform: translateY(-50%);
	& > svg {
		color: ${({ theme }) => theme.colors.secondary};
	}
`;

export const SelectDropdown = styled.div<{ open: boolean }>`
	position: fixed;
	top: 0;
	bottom: 0;
	right: ${({ open }) => (open ? "0" : "-100%")};
	width: 75%;
	background: ${({ theme }) => theme.colors.secondary};
	padding: 2rem;
	transition: 0.3s ease-out all;
	z-index: 5;

	& > h1 {
		font-size: 1.5rem;
		margin-bottom: 2.5rem;
	}
`;

export const SelectOption = styled.div<{ selected: boolean }>`
	display: flex;

	&:not(:last-child) {
		margin-bottom: 1rem;
	}

	& > svg {
		margin-right: 0.5rem;
		color: ${({ theme }) => theme.colors.primary};
	}

	& > span {
		font-weight: ${({ selected }) => selected && "bold"};
	}
`;
