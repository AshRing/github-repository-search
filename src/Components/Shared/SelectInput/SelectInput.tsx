import * as React from "react";

import {
	DrowdownArrow,
	SelectDropdown,
	SelectInputContainer,
	SelectLabel,
	SelectOption,
	SelectedText,
} from "./SelectInput.styles";
import { faArrowDown, faArrowUp, faCheckSquare, faSquare } from "@fortawesome/free-solid-svg-icons";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import useOnClickOutside from "use-onclickoutside";

interface Props {
	handleChange: (selected: string[]) => void;
	label: string;
	multiselect?: boolean;
	options: string[];
	selected: string[];
}

export const SelectInput = ({ handleChange, label, multiselect, options, selected }: Props) => {
	const [dropdownOpen, toggleDropdown] = React.useState<boolean>(false);
	const [multiselectSelectedOptions, setMultiselectSelectedOptions] =
		React.useState<string[]>(selected);
	const dropdownRef: React.RefObject<HTMLDivElement> = React.useRef();
	useOnClickOutside(dropdownRef, () => handleClickOutside());

	React.useEffect(() => {
		if (multiselect) {
			setMultiselectSelectedOptions(selected);
		}
	}, [selected]);

	const handleClickOutside = () => {
		if (multiselect) {
			if (JSON.stringify(multiselectSelectedOptions) !== JSON.stringify(selected)) {
				setMultiselectSelectedOptions(selected);
			}
		}
		toggleDropdown(false);
	};

	const handleMultiselectOptionSelection = (multiselectSelectedOption: string) => {
		if (multiselectSelectedOptions.includes(multiselectSelectedOption)) {
			const optionRemoved: string[] = multiselectSelectedOptions.filter(
				(selectedOption: string) => selectedOption !== multiselectSelectedOption,
			);
			setMultiselectSelectedOptions(optionRemoved);
		} else {
			setMultiselectSelectedOptions([
				...multiselectSelectedOptions,
				multiselectSelectedOption,
			]);
		}
	};

	const renderOptions = () => {
		return options.map((option: string, i: number) => {
			if (multiselect) {
				const optionSelected: boolean = multiselectSelectedOptions.includes(option);

				return (
					<SelectOption
						key={`${option}-${i}`}
						selected={optionSelected}
						onClick={() => handleMultiselectOptionSelection(option)}
					>
						{optionSelected ? (
							<FontAwesomeIcon icon={faCheckSquare} />
						) : (
							<FontAwesomeIcon icon={faSquare} />
						)}
						<span>{option}</span>
					</SelectOption>
				);
			} else {
				const optionSelected: boolean = selected.includes(option);
				return (
					<SelectOption
						key={`${option}-${i}`}
						selected={optionSelected}
						onClick={() => {
							handleChange([option]);
							toggleDropdown(false);
						}}
					>
						<span>{option}</span>
					</SelectOption>
				);
			}
		});
	};

	return (
		<SelectInputContainer
			inputActive={selected.length || dropdownOpen}
			onClick={() => (!dropdownOpen ? toggleDropdown(true) : undefined)}
		>
			<SelectLabel shrinkLabel={selected.length}>{label}</SelectLabel>
			<SelectedText>{selected.join(", ")}</SelectedText>
			<DrowdownArrow>
				{dropdownOpen ? (
					<FontAwesomeIcon icon={faArrowUp} />
				) : (
					<FontAwesomeIcon icon={faArrowDown} />
				)}
			</DrowdownArrow>
			<SelectDropdown open={dropdownOpen} ref={dropdownRef}>
				<h1>{label}</h1>
				{renderOptions()}
			</SelectDropdown>
		</SelectInputContainer>
	);
};
