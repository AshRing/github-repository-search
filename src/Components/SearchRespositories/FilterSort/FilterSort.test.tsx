import * as React from "react";

import { FilterSort, filters, sortInputs } from ".";
import { FilterSortContainer, SearchInput } from "./FilterSort.styles";
import { mount, shallow } from "enzyme";

import { IFilterSortOption } from "../../../_types";
import { SelectInput } from "./SelectInput";
import { TestWrapper } from "../../../_mocks";

let props;
beforeEach(() => {
	props = {
		addFilter: jest.fn(),
		changeSearchTerm: jest.fn(),
		changeSort: jest.fn(),
		filterBy: [],
		searchRepositories: jest.fn(),
		searchTerm: "",
		sortBy: sortInputs.bestMatch,
	};
});

test("FilterSort matches snapshot", () => {
	const wrapper = shallow(<FilterSort {...props} />);
	expect(wrapper).toMatchSnapshot();
});

test("changes search term on search input change", () => {
	const searchTerm = "test";
	const wrapper = mount(
		<TestWrapper>
			<FilterSort {...props} />
		</TestWrapper>,
	);
	wrapper.find(SearchInput).simulate("change", { target: { value: searchTerm } });
	expect(props.changeSearchTerm).toHaveBeenCalledWith(searchTerm);
});

test("doesn't display reset search button if term is empty", () => {
	props = {
		...props,
		searchTerm: "",
	};
	const wrapper = mount(
		<TestWrapper>
			<FilterSort {...props} />
		</TestWrapper>,
	);
	expect(wrapper.find("#resetSearchButton").exists()).toBeFalsy();
});

test("displays reset search button if term is populated & initiates reset", () => {
	props = {
		...props,
		searchTerm: "test",
	};
	const wrapper = mount(
		<TestWrapper>
			<FilterSort {...props} />
		</TestWrapper>,
	);
	expect(wrapper.find("#resetSearchButton").exists()).toBeTruthy();
	wrapper.find("#resetSearchButton").simulate("click");
	expect(props.changeSearchTerm).toHaveBeenCalledWith("");
});

test("language filter input handles filter change", () => {
	const selected = ["HTML"];
	const wrapper = mount(
		<TestWrapper>
			<FilterSort {...props} />
		</TestWrapper>,
	);
	wrapper.find("#languageFilterInput").find(SelectInput).prop("handleChange")(selected);
	expect(props.addFilter).toHaveBeenCalledWith({ name: filters.language, values: selected });
});

test("language filter selected prop is empty array if filter doesn't exist", () => {
	props = {
		...props,
		filterBy: [],
	};
	const wrapper = mount(
		<TestWrapper>
			<FilterSort {...props} />
		</TestWrapper>,
	);
	expect(wrapper.find("#languageFilterInput").find(SelectInput).prop("selected")).toEqual([]);
});

test("language filter selected prop contains language values if filter exists", () => {
	const languageFilter: IFilterSortOption = { name: filters.language, values: ["HTML"] };
	props = {
		...props,
		filterBy: [languageFilter],
	};
	const wrapper = mount(
		<TestWrapper>
			<FilterSort {...props} />
		</TestWrapper>,
	);
	expect(wrapper.find("#languageFilterInput").find(SelectInput).prop("selected")).toEqual(
		languageFilter.values,
	);
});

test("sort input handles change", () => {
	const selected = [sortInputs.stars];
	const wrapper = mount(
		<TestWrapper>
			<FilterSort {...props} />
		</TestWrapper>,
	);
	wrapper.find("#sortInput").find(SelectInput).prop("handleChange")(selected);
	expect(props.changeSort).toHaveBeenCalledWith(selected[0]);
});

test("submit searches repos", () => {
	const wrapper = mount(
		<TestWrapper>
			<FilterSort {...props} />
		</TestWrapper>,
	);
	wrapper.find(FilterSortContainer).simulate("submit");
	expect(props.searchRepositories).toHaveBeenCalled();
});
