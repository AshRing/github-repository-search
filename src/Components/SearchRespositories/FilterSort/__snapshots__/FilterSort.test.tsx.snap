// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterSort matches snapshot 1`] = `ShallowWrapper {}`;
